### ISSUE TITLE
Bug: Wrong calculation for parent & child products

### ISSUE THREAD

**Issue by Uatschitchun** [grocy/grocy#1430]
Steps to reproduce:

* Add product A & enable sum up children
* Add product B as child to A with 'factor purchase to stock quantity unit' being 0,5
* Make sure A & B use same QUs
* Buy maybe 2 items B
=> Result is A = 0 with a sum of 5

Idea behind was having a parent product being measured in cans and having 2 children (large & small cans) with small cans being 0,5 large cans (eg. canned tomatoes).

You could also set purchase factor of child to 0,25 and buy 4, which results in 17!?

_P.S. Sorry for spamming, but weather is bad and we're trying to setup grocy ;)_

**Comment by berrnd** [comment]
Please use a [private demo instance](https://www.reddit.com/r/grocy/comments/jwnk9b) to share such scenarios, also mentioned on the issue template - there are most of the time much more details to be considered than what you described...

I've set up your example here: https://test-klmwb18xo5e0v893pp6ceyq.demo-prerelease.grocy.info

Works, buying 5 of `Product B` results in a sum of 2.5 for `Product A`.

So please check what's different on your side.
If you maybe used the same QU for purchase _and_ stock + a factor, maybe that's the problem - factors need different units to work (and yes, the UI product edit page could enforce that).

The product option `Accumulate sub products min. stock amount´ has nothing to do with how _stock amounts_ of child products are aggregated into the parent products, it's only relevant for checking minimum stock amounts - like the name and the option hint implies.

**Comment by Uatschitchun** [comment]
> Please use a private demo instance

This was tested on the demo instance.

> If you maybe used the same QU for purchase and stock + a factor, maybe that's the problem

As it was written:

> Make sure A & B use same QUs

Yes, using same QU for both, stock & purchase

> If you maybe used the same QU for purchase and stock + a factor, maybe that's the problem - factors need different units to work (and yes, the UI product edit page could enforce that).

The UI doesn't prohibit setting both to the same QU with using a factor and it's also quite logical to do so. Stocking & purchasing beer in litres is possible in 0,33l, 0,5l, 1l (Tuborg) and 15l caskets. So it's needed to add QUs for small bottles, normal bottles, large bottles and caskets?
So because factoring the same QU isn't possible, I need to add QUs, like the above (small, normal, large & caskets) and even need to add small cans, cans, large cans (Metro). If I have other products in cans (eg. condensed milk), I need to also have small milk cans, medium milk cans and large milk cans. I also need to have litre, ml, cl, dl ...just because factoring the same QU isn't possible?!

> The product option `Accumulate sub products min. stock amount´ has nothing to do with how stock amounts of child products are aggregated into the parent products, it's only relevant for checking minimum stock amounts - like the name and the option hint implies.
Yes, clear so far. Thing is, there are different sizes of the same QU. Not wanting to clutter the shopping list with small, medium, large, etc, one could just sum up those various sizes possible to purchase under a relevant parent (beer, tomatoes, etc.) and also check "never show stock in overview" for the different children. This would result in only having beer, tomatoes, etc on the shopping list, setting a min. amount for the parent and factor the sizes up for to reach that min amount...?

**Comment by Uatschitchun** [comment]
https://test-yhgu19pq1xkt9x8r486lrh.demo.grocy.info/quantityunit/17
See Tonne & Zentner. 
Zentner: 1 Zentner = 5 Tonnen
Tonne: 1 Tonne = 20 Zentner
!?

**Comment by Uatschitchun** [comment]
Take another example:
Noodles: stock is gram and purchase is package with a factor of 500 (1 pkg = 500g)
Buy 1 pkg noodles => 500g. Ok.
Now set a unit conversion for package:
1 pkg = 1000g.
Now again buy 1 pkg noodles => 1000g. Wrong!

**Comment by Uatschitchun** [comment]
To make it short:
There need to be 2 separate factorings (conversions)! 
1.) Units: hl -> l -> dl -> cl -> ml -> EL -> TL (Tsp) 
2.) Packagings: Can, small can, medium can, etc. 
With 1.) being set in QU-conversions and 2.) being factors within the product.

**Comment by Uatschitchun** [comment]
And: if a program is allowing same units and a factor without complaining and is calculating wrong then, this is what I call a bug!
My math teacher would have marked 1+1=3 as failure and not just close it.

**Comment by berrnd** [comment]
The only thing that makes sense if you want to distinguish between Can sizes (0.33l, 0.5l and so on) is to add them as different QUs, or track the product in liters or something similar - everything else would require some guessing behind, at least when using the parent product in recipes.

And of course a conversion between two QUs can only be defined once per product.

> See Tonne & Zentner.
> Zentner: 1 Zentner = 5 Tonnen
> Tonne: 1 Tonne = 20 Zentner
> !?

Don't know what you want to tell me by that, can't see something wrong...

> Buy 1 pkg noodles => 500g. Ok.
> Now set a unit conversion for package:
> 1 pkg = 1000g.
> Now again buy 1 pkg noodles => 1000g. Wrong!

If you are referring to the mentioned private demo instance, that's not how it is currently configured there.
And that's then an indirect conversion (defined on QU level, not on product level), that's a duplicate of #1264.

> Units: hl -> l -> dl -> cl -> ml -> EL -> TL (Tsp)

Conversion on QU level are currently limited to 1 level (like parent/child products), duplicate of #1360

> should and could be hardcoded as SI (système international) as it's essential for meals

Duplicate of #912, however, at the end what you really want to use will always be something custom - so set it up once, done.

And sorry, maybe it's me, but it's really hard to follow your writing.
Please keep it to one topic and refer to something you set up on a private demo instance - everything else always misses some details and therefore is more than hard to reproduce, as we can see it here.

---

> And: if a program is allowing same units and a factor without complaining and is calculating wrong then, this is what I call a bug!
> My math teacher would have marked 1+1=3 as failure

I'm kind of allergic to such general lectures.
grocy is Open Source, everything can be fixed and improved by everyone, so far it was me for maybe about 99.9 %.
Any reproducible and precisely reported bug will be tracked here and hopefully fixed some time, the same goes for any feature request.
However, mixing so much into one request without a clear reference to a reproducible example (including all details, so, again, the only way is private demo instance per request/topic, nothing else makes sense, at least at a certain point of complexity of the request) is more than hard to follow.

> and not just close it.

Was initially only closed since I had no reproducible way, I reproduced it by what you provided and it worked, as it can be seen above.
In conclusion (still partly guessing due to missing details) it's most probably all about indirect QU conversions which is, as mentioned above, already reported in #1264 (a report with two precisely reported reproducible examples without any lectures about how math works).

So if you want that something get fixed or improved, please describe it by one request per topic and when it's a bug, please check existing ones before reporting.

This ends up being one of those requests that came up too often lately (mostly by private mails however) which eat a lot of time and lead to the fact that my desire to work on grocy shrinks. Some time ago I had fun to build grocy. Currently that's just not the case as it can be seen publicly here and it's not because there are bugs (in the past that was often my motivation and I fixed some of them kind of faster than they were reported, as it also can be seen here), it's just the way some people try to report them.

If you know how to do things just right, just do it.
If grocy has too many bugs for you, don't use it.
Plot twist: I still use grocy daily for myself and I'm personally not affected by any of the open bugs.
